source/opt/cfg_cleanup_pass.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a pass to cleanup the CFG to remove superfluous
+// constructs (e.g., unreachable basic blocks, empty control flow structures,
+// etc)
+
+#include <queue>
+#include <unordered_set>
+
+#include "cfg_cleanup_pass.h"
+
+#include "function.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+void CFGCleanupPass::RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                                  ir::Instruction* inst) {
+  uint32_t inst_id = inst->result_id();
+  if (inst_id == 0) {
+    return;
+  }
+
+  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
+  if (uses == nullptr) {
+    return;
+  }
+
+  for (auto u : *uses) {
+    if (u.inst->opcode() != SpvOpPhi) {
+      continue;
+    }
+
+    ir::Instruction* phi_inst = u.inst;
+    std::vector<ir::Operand> keep_operands;
+    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
+      const ir::Operand& var_op = phi_inst->GetOperand(i);
+      if (i >= 2 && i < phi_inst->NumOperands() - 1) {
+        // PHI arguments start at index 2. Each argument consists of two
+        // operands: the variable id and the originating block id.
+        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
+        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
+               "Phi operands should have exactly one word in them.");
+        uint32_t var_id = var_op.words.front();
+        uint32_t block_id = block_op.words.front();
+        if (var_id == inst_id && block_id == block.id()) {
+          i++;
+          continue;
+        }
+      }
+
+      keep_operands.push_back(var_op);
+    }
+
+    phi_inst->ReplaceOperands(keep_operands);
+  }
+}
+
+void CFGCleanupPass::RemoveBlock(ir::Function::iterator* bi) {
+  auto& block = **bi;
+  block.ForEachInst([&block, this](ir::Instruction* inst) {
+    // Note that we do not kill the block label instruction here. The label
+    // instruction is needed to identify the block, which is needed by the
+    // removal of PHI operands.
+    if (inst != block.GetLabelInst()) {
+      RemoveFromReachedPhiOperands(block, inst);
+      KillNamesAndDecorates(inst);
+      def_use_mgr_->KillInst(inst);
+    }
+  });
+
+  // Remove the label instruction last.
+  def_use_mgr_->KillInst(block.GetLabelInst());
```

Comment:
What if the label has a name?

This code will produce invalid SPIR-V when run on the following input:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main"
               OpSource GLSL 430
               OpName %main "main"
               OpName %dead "dead"
       %void = OpTypeVoid
          %5 = OpTypeFunction %void
       %main = OpFunction %void None %5
          %6 = OpLabel
               OpReturn
       %dead = OpLabel
               OpReturn
               OpFunctionEnd
```

%dead is dead, and will be swept away.  But then the OpName targeting %dead will be left around.

Should use KillNamesAndDecorates on the label instruction.

Please add the above as a test case too.